### ISSUE TITLE
Compass does not compile with Unix line-endings on Windows

### ISSUE THREAD

**Issue by jackie** [Compass/compass#949]
Compass version: 0.12.2
Ruby version: tested on 1.9.3 and 1.8.7
Windows version: Windows XP and Windows 7

Steps to recreate:
1. config.rb file in a compass project sets sass_options = {:unix_newlines => true}
2. on the command line in Windows, watch the project using compass watch and have the files recompile

What I expect to happen: the files get compiled with unix line endings.

What actually happens: the files compile with Windows line endings.

We can see the difference because the members on our team who are on a Mac compile and commit to the same SVN repo without issue, but when a Windows user watches the same set of files, they get re-compiled with no differences other than line endings.

**Comment by jackie** [comment]
I'm going to close this because after realizing that Sass was up to 3.1.20, the problem stopped happening. (We were all on 3.1.19 when I filed the ticket.)

**Comment by olegtaranenko** [comment]
-On my windows box I've done:

**gem update sass**
Updating installed gems
Updating sass
Fetching: sass-3.1.20.gem (100%)
Successfully installed sass-3.1.20
Gems updated: sass
Installing ri documentation for sass-3.1.20...
Installing RDoc documentation for sass-3.1.20...

**get update compass**
Updating installed gems
Updating compass
Fetching: compass-0.12.2.gem (100%)
Successfully installed compass-0.12.2
Gems updated: compass
Installing ri documentation for compass-0.12.2...
Installing RDoc documentation for compass-0.12.2...

**config.rb**

# Get the directory that this configuration file exists in
dir = File.dirname(**FILE**)

# Load the sencha-touch framework automatically.
load File.join(dir, '..', '..', 'sdk', 'resources', 'themes')

# Compass configurations
sass_path = dir
css_path = File.join(dir, "..", "css")

 # Require any additional compass plugins here.
images_dir = File.join(dir, "..", "images")
output_style = :expanded
environment = :production

sass_options = {:unix_newlines => true}

Still no joy --- still windows line ending :@...
may be you change some more options?
thanks

**Comment by jtmohr** [comment]
I'm still not able to get Compass to compile files with Unix line endings on Windows either.

Steps to reproduce on Windows 7 Pro:
1. > compass create compass-test
2. Add: sass_options = {:unix_newlines => true} to project's config.rb file
3. Add some css to screen.css
4. > compass compile compass-test
5. screen.css is overwritten
6. screen.css still has Windows line endings

Is it possible to Reopen this issue or should we create a new issue?

**Comment by jackie** [comment]
I'm reopening because we are also still having this issue, even after updating to 3.1.20 (I know I said updating previously solved the issue, not sure what happened, but we are still having the problem, even with Sass/Comapss up to date).

**Comment by SebTardif** [comment]
A new test about this would help not having regression.

**Comment by scottdavis** [comment]
Should probably post this on the sass issues list. Compass doesn't do the actual compile, sass does.

**Comment by etramell** [comment]
I am seeing this on Windows 7, with Compass v0.12.2 and Sass v3.2.8.  I don't know if this is true, but it seems to me that setting the "sass_options = {:unix_newlines => true}" in Compass does no good because sass doesn't actually write (close) the file?

From http://sass-lang.com/docs/yardoc/file.SASS_REFERENCE.html:
:unix_newlines
If true, use Unix-style newlines when writing files. Only has meaning on Windows, and only when Sass is writing the files (in Rack, Rails, or Merb, when using Sass::Plugin directly, or when using the command-line executable).

Does there need to be a specific Compass setting for this instead?

**Comment by materliu** [comment]
of course, it does not work, can anyone find a way to solve this problem ?

**Comment by johnmave126** [comment]
I did some investigation in the code and found some interesting points.
**lib/compass/compiler.rb:136**

``` ruby
    # Compile one Sass file
    def compile(sass_filename, css_filename)
      start_time = end_time = nil
      css_content = logger.red do
        timed do
          engine(sass_filename, css_filename).render
        end
      end
      duration = options[:time] ? "(#{(css_content.__duration * 1000).round / 1000.0}s)" : ""
      write_file(css_filename, css_content, options.merge(:force => true, :extra => duration))
      Compass.configuration.run_stylesheet_saved(css_filename)
    end
```

The file is written by

``` ruby
      write_file(css_filename, css_content, options.merge(:force => true, :extra => duration))
```

**lib/compass/actions.rb:52**

``` ruby
      if skip_write
        FileUtils.touch file_name unless options[:dry_run]
      else
        mode = "w"
        mode << "b" if binary
        open(file_name, mode) do |file|
          file.write(contents)
        end
```

The key point is that by default the file is not written in binary mode. On windows, the line break will be substitute to CR/LF instead of LF generated by sass.
I have a very primitive patch below

``` diff
diff --git a/lib/compass/compiler.rb b/lib/compass/compiler.rb
--- a/lib/compass/compiler.rb
+++ b/lib/compass/compiler.rb
@@ -11,6 +11,7 @@
      self.logger = options.delete(:logger)
      sass_opts = options.delete(:sass) || {}
      self.options = options
+      self.options[:unix_newlines] = sass_opts[:unix_newlines]
      self.sass_options = options.dup
      self.sass_options.delete(:quiet)
      self.sass_options.update(sass_opts)
diff --git a/lib/compass/actions.rb b/lib/compass/actions.rb
--- a/lib/compass/actions.rb
+++ b/lib/compass/actions.rb
@@ -53,7 +53,7 @@
        FileUtils.touch file_name unless options[:dry_run]
      else
        mode = "w"
-        mode << "b" if binary
+        mode << "b" if binary or options[:unix_newlines]
        open(file_name, mode) do |file|
          file.write(contents)
        end
```

The patch is based on compass 0.12.2, and it is very primitive. I am not quite familiar with ruby and compass, so I don't know the use case of binary flag of write_file and the options management in compass. This patch is a very **dirty** one, and I hope some one who is familiar with this project can improve it.

**Comment by chriseppstein** [comment]
This is fixed on master for 1.0.
